// @stepzen/sdk/lib/commands/getPublicAccount.js
"use strict";
// Copyright IBM Corp. 2020, 2024
Object.defineProperty(exports, "__esModule", { value: true });
const debug = require("debug");
const fetch_1 = require("@stepzen/fetch");
const request_1 = require("../shared/request");
exports.default = async (options, sdkConfig) => {
    const account = process.env.STEPZEN_PUBLIC_ACCOUNT_API_ACCOUNT || 'stepzen';
    const endpoint = process.env.STEPZEN_PUBLIC_ACCOUNT_API_ENDPOINT || 'api/publicaccount';
    const server = options.server
        .replace('{account}', account)
        .replace('.io', '.net');
    const url = new URL(`${server}/${endpoint}/__graphql`);
    // Inlclude the token into the URL so that it is visible in the logs
    // (allows StepZen to do analytics based on the GCP logs).
    url.searchParams.set('token', options.publicAccountToken);
    try {
        debug('stepzen:createAnonymousAccount')(url);
        const response = await (0, fetch_1.default)(url, {
            method: 'POST',
            headers: {
                'Content-Type': 'application/json',
                'User-Agent': (0, request_1.getUserAgent)(sdkConfig),
            },
            body: JSON.stringify({
                query: `query (
            $token: String!
          ) {
            getAccountDetails(
              token: $token
            ) {
              account: accountName
              adminkey: adminKey
              apikey: apiKey
            }
          }`,
                variables: {
                    token: options.publicAccountToken,
                },
            }),
        });
        const json = await response.json();
        debug('stepzen:createAnonymousAccount')(json);
        if (json.errors) {
            return {
                success: false,
                errors: json.errors,
            };
        }
        if (!json.data?.getAccountDetails) {
            throw new Error('No data returned from the API endpoint.');
        }
        return json.data?.getAccountDetails;
    }
    catch (error) {
        throw new Error(`Could not create a public account (${error})`);
    }
};
