// @stepzen/sdk/lib/commands/upload.js
"use strict";
// Copyright IBM Corp. 2020, 2024
Object.defineProperty(exports, "__esModule", { value: true });
const debug = require("debug");
const path = require("path");
const fetch_1 = require("@stepzen/fetch");
const constants_1 = require("../shared/constants");
const payloads_1 = require("../shared/payloads");
const request_1 = require("../shared/request");
const transpiling_1 = require("../shared/transpiling");
const validation_1 = require("../shared/validation");
const rmtemp_1 = require("../shared/rmtemp");
exports.default = async (details, account, sdkConfig) => {
    const headers = (0, request_1.getRequestHeaders)(account, sdkConfig);
    let payload;
    switch (details.type) {
        case 'configurationset':
            await (0, validation_1.validateConfigurationset)(details.file);
            const transpiled = await (0, transpiling_1.transpileConfigurationset)(details.file);
            try {
                debug('stepzen:sdk')(`effective config written to ${transpiled}`);
                payload = await (0, payloads_1.generateYamlPayload)(transpiled);
            }
            finally {
                (0, rmtemp_1.rmtemp)(path.dirname(transpiled));
            }
            break;
        case 'schema':
            await (0, validation_1.validateSchema)(details.directory);
            payload = await (0, payloads_1.generateZipPayload)(details.directory, {
                destination: details.destination,
                stitcherName: 'index.graphql',
            }, [/.*\.graphql$/i]);
            break;
    }
    debug('stepzen:headers')(headers);
    debug('stepzen:payload')(payload);
    const response = await (0, fetch_1.default)(`${account.server}${constants_1.ADMIN_UPLOAD_URL}/${details.type}/${details.destination}`, {
        body: payload,
        headers: headers,
        method: 'POST',
    });
    debug('stepzen:response')(response);
    if (response.status >= 400 && response.status < 500) {
        return {
            message: 'Could not complete the request. Please check your authentication details are correct.',
            success: false,
        };
    }
    if (response.status >= 500) {
        return {
            errors: ['Internal Server Error'],
            message: response.statusText,
            success: false,
        };
    }
    const text = await response.text();
    let json;
    try {
        json = JSON.parse(text);
    }
    catch {
        return {
            errors: [`An unexpected error occurred.\n\n${text}`],
            message: `An unexpected error occurred.\n\n${text}`,
            success: false,
        };
    }
    return json;
};
