// stepzen/lib/commands/init.js
"use strict";
// Copyright IBM Corp. 2020, 2024
Object.defineProperty(exports, "__esModule", { value: true });
const core_1 = require("@oclif/core");
const workspace_1 = require("../shared/workspace");
const zen_command_1 = require("../shared/zen-command");
class Init extends zen_command_1.default {
    async run() {
        const { args, flags } = await this.parse(Init);
        const created = await (0, workspace_1.initWorkspace)({
            directory: args.directory,
            endpoint: flags.endpoint,
            yes: flags.yes,
            nonInteractive: flags['non-interactive'],
        });
        this.log(`Created a StepZen workspace in ${created.directory}`);
        return created;
    }
}
exports.default = Init;
Init.description = 'Initialize a StepZen workspace in the current directory.';
Init.flags = {
    ...zen_command_1.default.flags,
    endpoint: core_1.Flags.string({
        description: 'Endpoint name (e.g. api/myapp)',
    }),
    yes: core_1.Flags.boolean({
        description: 'Accept default answers to interactive prompts',
        default: false,
    }),
};
Init.args = {
    directory: core_1.Args.string({ hidden: true }),
};
