// stepzen/lib/commands/import/snowflake.js
"use strict";
// Copyright IBM Corp. 2020, 2024
Object.defineProperty(exports, "__esModule", { value: true });
const core_1 = require("@oclif/core");
const errors_1 = require("@oclif/core/lib/errors");
const chalk = require("chalk");
const lodash_1 = require("lodash");
const flags_1 = require("../../generate/flags");
const import_command_1 = require("../../generate/import-command");
const questions_1 = require("../../generate/questions");
const sql2sdl_1 = require("../../generate/sql2sdl");
const configuration_1 = require("../../shared/configuration");
const dsn_parser_1 = require("@soluble/dsn-parser");
const inquirer = require("../../shared/inquirer");
const utils_1 = require("../../shared/utils");
class ImportSnowflake extends import_command_1.ImportCommandBase {
    get source() {
        return 'snowflake';
    }
    async import(commonOptions) {
        const { args, flags } = await this.parse(ImportSnowflake);
        const nonInteractiveOptions = {
            ...(args.dsn ? this.getImportOptionsFromDsn(args.dsn) : {}),
            ...this.getImportOptionsFromFlags(flags),
        };
        const requiredOptions = [
            'accountId',
            'user',
            'password',
            'database',
            'warehouse',
        ];
        const hasRequiredOptions = requiredOptions.every(option => nonInteractiveOptions[option] !== undefined);
        let sql2sdlOptions;
        if (hasRequiredOptions) {
            // an ugly hint for the TS compiler because it is not smart enough to figure this :(
            const typedNonInteractiveOptions = nonInteractiveOptions;
            // run non-interactively
            sql2sdlOptions = {
                dbType: this.source,
                ...commonOptions,
                ...typedNonInteractiveOptions,
            };
        }
        else {
            if (flags['non-interactive']) {
                throw new errors_1.CLIError('Please provide a Data Source Name (DSN) string for a Snowflake database.' +
                    chalk `\nSee more help with {bold stepzen import snowflake --help}`);
            }
            // no parameters given: start an interactive prompt
            if (!flags.silent) {
                this.log();
                this.log(chalk `{bold stepzen import snowflake} - ` +
                    chalk.dim('introspect a Snowflake database and extend your GraphQL schema' +
                        ' with the types, queries and mutations for accessing it through' +
                        ' a StepZen API.'));
                this.log();
            }
            // one or more required parameters are missing: start an interactive prompt
            const interactiveOptions = await this.getImportOptionsInteractively(nonInteractiveOptions);
            sql2sdlOptions = {
                dbType: this.source,
                ...commonOptions,
                // Some options that may have been provided via the command line are
                // excluded from the interactive prompts (e.g. `--db-include`). That's
                // why the non-interactive options are also included here.
                ...nonInteractiveOptions,
                ...interactiveOptions,
            };
        }
        // Snowflake is very particular about the letter case in the database and schema names
        sql2sdlOptions.database = sql2sdlOptions.database.toUpperCase();
        sql2sdlOptions.schema = sql2sdlOptions.schema
            ? sql2sdlOptions.schema.toUpperCase()
            : 'PUBLIC';
        const configuration = await (0, configuration_1.readConfiguration)();
        return this.wrapInProgressAndTimeout((0, sql2sdl_1.sql2sdl)(sql2sdlOptions, configuration));
    }
    getImportOptionsFromFlags(flags) {
        const options = {
            accountId: flags['snowflake-account-id'],
            user: flags['db-user'],
            password: flags['db-password'],
            database: flags['db-database'],
            schema: flags['db-schema'],
            warehouse: flags['snowflake-warehouse'],
            linkTypes: flags['db-link-types'],
            include: flags['db-include'],
            introspectionOptions: {
                // Use the 2023 naming convention by default, and the 2022 naming
                // convention only if explicitly requested by the user.
                naming: flags['db-use-deprecated-2022-naming'] ? '2022' : '2023',
            },
        };
        // Remove `undefined`-valued properties
        return (0, lodash_1.pickBy)(options, value => value !== undefined);
    }
    getImportOptionsFromDsn(possibleDsn) {
        const parsed = (0, dsn_parser_1.parseDsn)(possibleDsn);
        if (parsed.success) {
            const value = parsed.value;
            const options = {
                accountId: value.host,
                user: value.user ?? value.params?.user,
                password: value.pass ?? value.params?.password,
                database: value.db,
                schema: value.params?.schema === undefined
                    ? undefined
                    : `${value.params?.schema}`,
                warehouse: value.params?.warehouse === undefined
                    ? undefined
                    : `${value.params?.warehouse}`,
            };
            // Remove `undefined`-valued properties
            return (0, lodash_1.pickBy)(options, value => value !== undefined);
        }
        throw new errors_1.CLIError(parsed.message);
    }
    async getImportOptionsInteractively(defaults = {}) {
        const questions = inquirer.overrideDefaults([
            {
                name: 'accountId',
                message: 'What is your Snowflake account identifier?',
                validate: input => {
                    const { error } = (0, utils_1.parseSnowflakeAccountId)(input);
                    return error ? error.message : true;
                },
            },
            questions_1.ImportQuestions.dbUser(),
            questions_1.ImportQuestions.dbPassword(),
            questions_1.ImportQuestions.dbDatabase(),
            {
                ...questions_1.ImportQuestions.dbSchema(),
                filter: input => input.toUpperCase(),
            },
            {
                name: 'warehouse',
                message: 'What is your Snowflake warehouse?',
                validate: input => input.trim() !== '' || 'Snowflake warehouse must not be empty',
            },
            questions_1.ImportQuestions.dbLinkTypes(),
        ], defaults);
        return inquirer.prompt(`import-${this.source}`, questions);
    }
}
exports.default = ImportSnowflake;
ImportSnowflake.description = `Import a schema for a Snowflake data source into your GraphQL API.` +
    `\n` +
    chalk `\n{bold stepzen import snowflake} automatically introspects a` +
    ` Snowflake database, generates a GraphQL schema for accessing this` +
    ` database through a StepZen API, and merges the generated types,` +
    ` queries and mutations into your GraphQL schema.`;
ImportSnowflake.flags = {
    ...import_command_1.ImportCommandBase.flags,
    // 'db-host': ImportFlags.dbHost(), -- intentionally omitted (Snowflake uses an "account ID" instead)
    'db-user': flags_1.ImportFlags.dbUser(),
    'db-password': flags_1.ImportFlags.dbPassword(),
    'db-database': flags_1.ImportFlags.dbDatabase(),
    'db-schema': flags_1.ImportFlags.dbSchema({ default: 'PUBLIC' }),
    'db-link-types': flags_1.ImportFlags.dbLinkTypes(),
    'db-include': flags_1.ImportFlags.dbInclude(),
    'db-use-deprecated-2022-naming': flags_1.ImportFlags.dbUseDeprecatedNaming(),
    'snowflake-account-id': core_1.Flags.string({
        description: chalk `Snowflake account identifier in the {bold orgname-accountname}` +
            chalk ` format. For more information, see the Snowflake documentation at` +
            chalk ` https://docs.snowflake.com/en/user-guide/admin-account-identifier.html.`,
    }),
    'snowflake-warehouse': core_1.Flags.string({
        description: 'Snowflake warehouse',
    }),
};
ImportSnowflake.args = {
    dsn: core_1.Args.string({
        description: chalk `{italic (optional)} Data Source Name (DSN) of a Snowflake database.` +
            chalk `\nExample: {bold snowflake://user:password@orgname-accountname/database?warehouse=warehouse&schema=schema}` +
            chalk `\n` +
            chalk `\nFlags, such as {bold --db-host}, override the corresponding` +
            chalk ` parts of the DSN (if both are provided).`,
    }),
};
