// stepzen/lib/generate/graphql2sdl.js
"use strict";
// Copyright IBM Corp. 2020, 2024
Object.defineProperty(exports, "__esModule", { value: true });
exports.graphql2sdl = void 0;
const constants_1 = require("../shared/constants");
const helpers_1 = require("./helpers");
const graphql2sdl = async ({ endpoint, typePrefix, includeRootOperations, headers }, configuration) => {
    return (0, helpers_1.queryIntrospectionService)((0, constants_1.getIntrospectionUrl)(configuration), {
        operation: 'getSDLFromGraphQL',
        variables: {
            endpoint: {
                type: 'String!',
                value: endpoint,
            },
            typePrefix: {
                type: 'String',
                value: typePrefix || null,
            },
            includeRootOperations: {
                type: 'Boolean',
                value: includeRootOperations || null,
            },
            headers: {
                type: '[HeaderInput!]',
                value: headers && headers.length > 0 ? headers : null,
            },
        },
    });
};
exports.graphql2sdl = graphql2sdl;
