// stepzen/lib/shared/curl-parser.js
"use strict";
// Copyright IBM Corp. 2020, 2024
Object.defineProperty(exports, "__esModule", { value: true });
exports.parseCurlArgv = void 0;
const constants_1 = require("./constants");
const header_1 = require("./header");
// Based on the existing curl flags at https://curl.se/docs/manpage.html
const curlFlagRegex = /^(-[a-zA-Z0-9.:#])|(--[a-zA-Z0-9.-]+)$/;
const httpURLRegex = /^https?:\/\//i;
const isAFlagArg = (arg) => curlFlagRegex.test(arg);
const isAURL = (arg) => httpURLRegex.test(arg);
const parseHeaderFlag = (value, partialCurlArgs) => {
    const { result: header, error } = (0, header_1.parseHeader)(value);
    if (error) {
        return { error };
    }
    const headers = partialCurlArgs.headers || [];
    // A `null` from parseCurlHeaderString() means a header should NOT be
    // sent. This is not supported by zenserv / the introspection service
    // so the CLI simply omits such headers
    if (header) {
        headers.push(header);
    }
    return {
        result: {
            ...partialCurlArgs,
            headers,
        },
    };
};
const parseDataFlag = (value, partialCurlArgs, match) => {
    if (match !== '--data-raw' && value.charAt(0) === '@') {
        return {
            error: {
                message: `Reading request data from local files in not currently supported ` +
                    `by StepZen CLI (${match} ${value}). If this is a blocker ` +
                    `for you, please let us know on Discord (${constants_1.STEPZEN_DISCORD_URL})`,
            },
        };
    }
    const data = (partialCurlArgs.data ? partialCurlArgs.data + '&' : '') + value;
    return {
        result: {
            ...partialCurlArgs,
            data,
        },
    };
};
const parseMethodFlag = (value, partialCurlArgs) => {
    const lowercaseMethod = value.toLowerCase();
    if (lowercaseMethod !== 'post' && lowercaseMethod !== 'get') {
        return {
            error: { message: `The method ${value} is currently not supported.` },
        };
    }
    const method = lowercaseMethod === 'post' ? 'Post' : 'Get';
    return {
        result: {
            ...partialCurlArgs,
            method,
        },
    };
};
const parseURLFlag = (value, partialCurlArgs) => {
    if (partialCurlArgs.url) {
        return {
            error: {
                message: `Multiple URLs are not currently supported by StepZen CLI (${value}).` +
                    ` If this is a blocker for you, please let us know on Discord (${constants_1.STEPZEN_DISCORD_URL})`,
            },
        };
    }
    return {
        result: {
            ...partialCurlArgs,
            url: value,
        },
    };
};
const flags = [
    {
        matches: ['-H', '--header'],
        parse: parseHeaderFlag,
    },
    {
        matches: ['-d', '--data', '--data-ascii', '--data-raw', '--data-binary'],
        parse: parseDataFlag,
    },
    {
        matches: ['-X', '--request'],
        parse: parseMethodFlag,
    },
    {
        matches: ['--url'],
        parse: parseURLFlag,
    },
];
const tryMatchCurlFlag = (matches, argv, i) => {
    for (const match of matches) {
        const isShortFlag = match.length === 2;
        if (isShortFlag && argv[i].startsWith(match) && argv[i].length > 2) {
            return {
                result: {
                    value: argv[i].substring(2),
                    match,
                },
            };
        }
        if (argv[i] === match) {
            if (i + 1 >= argv.length) {
                return {
                    error: { message: `The '${argv[i]}' curl flag requires a value` },
                };
            }
            return {
                result: {
                    value: argv[i + 1],
                    match,
                    skip: 1,
                },
            };
        }
    }
    return { result: undefined };
};
/**
 * Parse a curl command line arguments array to a JSON structure consumable by
 * the StepZen introspection service backend.
 *
 * @param {*} argv an array of curl arguments `[
 *      '-H',
 *      'api-key: my-secret-key',
 *      '--header',
 *      'x-custom-options: my-custom-option-2',
 *      'https://test.stepzen.net/version',
 *      '--dir',
 *      '/my-app'
 *    ]`
 * @returns {*} a structured object with the curl arguments
 */
const parseCurlArgv = (argv) => {
    if (argv[0]?.toLowerCase() === 'curl') {
        argv = argv.slice(1);
    }
    let result = {};
    for (let i = 0; i < argv.length; i++) {
        if (isAFlagArg(argv[i])) {
            let isKnownFlag = false;
            for (const flag of flags) {
                const { result: match, error: matchError } = tryMatchCurlFlag(flag.matches, argv, i);
                if (matchError) {
                    // flag matched but it requires a value which is missing
                    return { error: matchError };
                }
                if (!match) {
                    // no match => try matching the next flag
                    continue;
                }
                const { result: parseResult, error: parseError } = flag.parse(match.value, result, match.match);
                if (parseError) {
                    return { error: parseError };
                }
                result = parseResult;
                i += match.skip || 0;
                isKnownFlag = true;
                break;
            }
            if (!isKnownFlag) {
                return {
                    error: {
                        message: `The '${argv[i]}' curl flag is not currently supported by StepZen CLI.` +
                            ` If this is a blocker for you, please let us know on Discord (${constants_1.STEPZEN_DISCORD_URL})`,
                    },
                };
            }
        }
        else {
            // The only curl argument that's not a flag is the URL
            const parseResult = parseURLFlag(argv[i], result, '');
            if (parseResult.error) {
                return parseResult;
            }
            result = parseResult.result;
        }
    }
    if (!result.url) {
        return {
            error: { message: 'curl: a URL is required' },
        };
    }
    if (!isAURL(result.url)) {
        return {
            error: {
                message: `Unsupported URL schema in '${result.url}'.` +
                    ` StepZen CLI currently supports 'https' and 'http'.` +
                    ` If this is a blocker for you, please let us know on Discord (${constants_1.STEPZEN_DISCORD_URL})`,
            },
        };
    }
    result.headers = result.headers || [];
    result.method = result.method || (result.data ? 'Post' : 'Get');
    // Add the default content-type header if the request has any data
    // in it, and no content-type header is explicitly provided.
    if (result.headers.findIndex(header => header.name.toLowerCase() === 'content-type') === -1 &&
        result.data) {
        result.headers.push({
            name: 'content-type',
            value: 'application/x-www-form-urlencoded',
        });
    }
    return { result: result };
};
exports.parseCurlArgv = parseCurlArgv;
